locator/network_topology_strategy.cc

```diff
@@ -350,38 +351,135 @@ future<tablet_map> network_topology_strategy::reallocate_tablets(schema_ptr s, t
     co_return tablets;
 }
 
-future<tablet_replica_set> network_topology_strategy::reallocate_tablets(schema_ptr s, token_metadata_ptr tm, load_sketch& load, const tablet_map& cur_tablets, tablet_id tb) const {
+future<tablet_replica_set> network_topology_strategy::reallocate_tablets(schema_ptr s,
+    token_metadata_ptr tm,
+    load_sketch& load,
+    const tablet_map& cur_tablets,
+    tablet_id tb) const
+{
     tablet_replica_set replicas;
     // Current number of replicas per dc
     std::unordered_map<sstring, size_t> nodes_per_dc;
     // Current replicas per dc/rack
     std::unordered_map<sstring, std::map<sstring, std::unordered_set<locator::host_id>>> replicas_per_dc_rack;
+    std::unordered_map<sstring, rack_list> old_racks_per_dc;
 
     replicas = cur_tablets.get_tablet_info(tb).replicas;
     for (const auto& tr : replicas) {
         const auto& node = tm->get_topology().get_node(tr.host);
         replicas_per_dc_rack[node.dc_rack().dc][node.dc_rack().rack].insert(tr.host);
         ++nodes_per_dc[node.dc_rack().dc];
+        old_racks_per_dc[node.dc_rack().dc].push_back(node.dc_rack().rack);
     }
 
     // #22688 - take all dcs in topology into account when determining migration.
     // Any change should still have been pre-checked to never exceed rf factor one.
     for (const auto& dc : tm->get_topology().get_datacenters()) {
-        auto dc_rf = get_replication_factor(dc);
-        auto dc_node_count = nodes_per_dc[dc];
-        if (dc_rf == dc_node_count) {
-            continue;
-        }
-        if (dc_rf > dc_node_count) {
-            replicas = co_await add_tablets_in_dc(s, tm, load, tb, replicas_per_dc_rack[dc], replicas, dc, dc_node_count, dc_rf);
+        auto new_rf = get_replication_factor_data(dc);
+
+        if (new_rf && new_rf->is_rack_based()) {
+            auto diff = diff_racks(old_racks_per_dc[dc], new_rf->get_rack_list());
+
+            tablet_logger.debug("reallocate_tablets {}.{} tablet_id={} dc={} old_racks={} add_racks={} del_racks={}",
+                    s->ks_name(), s->cf_name(), tb, dc, old_racks_per_dc[dc], diff.added, diff.removed);
+
+            if (!diff) {
+                continue;
+            }
+
+            if (!diff.added.empty() && !diff.removed.empty()) {
+                // FIXME: Support replace
+                throw std::runtime_error("replacing racks unsupported");
+            } else if (!diff.added.empty()) {
+                replicas = add_tablets_in_racks(s, tm, load, tb, replicas, dc, diff.added);
+            } else { // diff.removed
+                replicas = drop_tablets_in_racks(s, tm, load, tb, replicas, dc, diff.removed);
```

Comment:
Can't we have both added and dropped racks?

Reply:
That would be rack replacement, which is not supported yet. 

If the removed rack is dead, it can be emulated. All nodes in it should be first marked as dead with removenode. Then you drop that rack from the RF list, then you add a new rack to the RF list.